File: examples/05-fork-join/Example4.java
// =================================================================
//
// File: Example4.java
// Authors:
//				 A01651517 Peter Glez
//				 A01703947 Juan Alcántara
// Description: This file contains the code to count the number of
//				even numbers within an array using Java's Fork-Join.
//
// Copyright (c) 2020 by Tecnologico de Monterrey.
// All Rights Reserved. May be reproduced for any non-commercial
// purpose.
//
// =================================================================
import java.util.concurrent.RecursiveTask;
import java.util.concurrent.ForkJoinPool;

public class Example4 extends RecursiveTask<Integer> {
	private static final int SIZE = 100_000_000;
	private static final int MIN = 100_000;

	private int array[];
	private int start;
	private int end;

	public Example4 (int array[], int start, int end) {
		this.array = array;
		this.start = start;
		this.end = end;

	}

	private Integer computeDirectly() {
		int result = 0;
		for (int i = start; i < end; i++) {
			if (array[i] % 2 == 0) 
				result++;
		}
		return result;
	}

	@Override
	protected Integer compute() {
		if (end - start <= MIN) {
			return computeDirectly();
		} else {
		  int mid = (start + end)/2;
			Example4 lower = new Example4(array, start, mid);
			lower.fork();
			Example4 high = new Example4(array, mid, end); 
			return high.compute() + lower.join();
		}
	}

	public static void main(String args[]) {
		long startTime, stopTime, result = 0;
		int array[];
		double ms;
		ForkJoinPool pool;

		array = new int[SIZE];
		Utils.fillArray(array);
		Utils.displayArray("array", array);

		System.out.printf("Starting with %d threads...\n", Utils.MAXTHREADS);
		ms = 0;
		for (int i = 0; i < Utils.N; i++) {
			startTime = System.currentTimeMillis();

			pool = new ForkJoinPool(Utils.MAXTHREADS);
			result = pool.invoke(new Example4(array, 0, array.length));

			stopTime = System.currentTimeMillis();
			ms += (stopTime - startTime);
		}
		System.out.printf("sum = %d\n", result);
		System.out.printf("avg time = %.5f ms\n", (ms / Utils.N));
	}

}

File: examples/07-tbb/example4.cpp
// =================================================================
//
<<<<<<< HEAD
// File: Example4.cpp
=======
// File: example4.cpp
>>>>>>> 36de68c7e4ca1f4f269371a62af84718703fdbaa
// Authors:
//				 A01651517 Pedro González
//				 A01703947 Juan Alcántara
// Description: This file contains the code to count the number of
//				even numbers within an array using Intel's TBB.
//              To compile: g++ example4.cpp -ltbb
//
// Copyright (c) 2020 by Tecnologico de Monterrey.
// All Rights Reserved. May be reproduced for any non-commercial
// purpose.
//
// =================================================================

#include <iostream>
#include <iomanip>
#include <cmath>
#include <tbb/parallel_reduce.h>
#include <tbb/blocked_range.h>
#include "utils.h"

#define SIZE 1000000000 //1e9

using namespace std;
using namespace tbb;

class CountEven {
	private:
		int *array, result;
	public:
		CountEven(int *arr) : array(arr), result(0) {}

		CountEven(CountEven &x, split) : array(x.array), result(0) {}

		int getResult() const {
			return result;
		}

		void operator() (const blocked_range<int> &r) {
			for (int i = r.begin(); i != r.end(); i++) {
				result += !(array[i] % 2);
			}
		}

		void join(const CountEven &x) {
			result += x.result;
		}
};

int main(int argc, char* argv[]) {
	int *a, result;
	double ms;

	a = new int[SIZE];
	random_array(a, SIZE);
	display_array("a", a);

	srand(time(0));

	cout << "Starting..." << endl;
	ms = 0;

	for (int i = 0; i < N; i++) {
		start_timer();

		CountEven counter(a);
		parallel_reduce(blocked_range<int>(0, SIZE), counter);
		result = counter.getResult();

		ms += stop_timer();
	}
	cout << "result = " << result << endl;
	cout << "avg time = " << setprecision(15) << (ms / N) << " ms" << endl;

	delete [] a;
	return 0;
}

